File: Source/OldCode/Archive/2005 TVA Code Library/Version/2.0/Local/Source/Tva.DatAWare/ArchiveDataBlock.vb
' 02/24/2007

Imports System.IO

Public Class ArchiveDataBlock

    Private m_location As Long
    Private m_size As Integer
    Private m_fileStream As FileStream
    Private m_writeCursor As Long

    Public Sub New(ByVal archiveFileStream As FileStream, ByVal location As Long, ByVal size As Integer)

        MyBase.New()
        m_fileStream = archiveFileStream
        m_location = location
        m_size = size
        m_writeCursor = location    ' This is where we'll start writing data to begin with.

    End Sub

    Public ReadOnly Property Location() As Long
        Get
            Return m_location
        End Get
    End Property

    ''' <summary>
    ''' 
    ''' </summary>
    ''' <value></value>
    ''' <returns></returns>
    ''' <remarks>In KB</remarks>
    Public ReadOnly Property Size() As Integer
        Get
            Return m_size
        End Get
    End Property

    Public ReadOnly Property Capacity() As Integer
        Get
            Return (m_size * 1024) \ StandardPointData.BinaryLength
        End Get
    End Property

    Public ReadOnly Property SlotsUsed() As Integer
        Get
            Return Convert.ToInt32((m_writeCursor - m_location) \ StandardPointData.BinaryLength)
        End Get
    End Property

    Public ReadOnly Property SlotsAvailable() As Integer
        Get
            Return Me.Capacity - Me.SlotsUsed
        End Get
    End Property

    Public Function Read() As List(Of StandardPointData)

        Dim data As New List(Of StandardPointData)()

        ' We'll start reading from where the data block begins.
        m_fileStream.Seek(m_location, SeekOrigin.Begin)

        Dim binaryImage As Byte() = CreateArray(Of Byte)(StandardPointData.BinaryLength)
        For i As Integer = 1 To Me.Capacity
            ' Read the binary data from the file and create StandardPointData instance form it.
            m_fileStream.Read(binaryImage, 0, binaryImage.Length)
            Dim pointData As New StandardPointData(binaryImage)
            If Not pointData.IsNull Then
                data.Add(pointData)
            Else
                ' The data we just read is blank so we'll roll-back the cursor and stop reading further.
                m_fileStream.Seek(-binaryImage.Length, SeekOrigin.Current)
                Exit For
            End If
        Next
        If m_writeCursor <> m_fileStream.Position Then m_writeCursor = m_fileStream.Position

        Return data

    End Function

    Public Sub Write(ByVal pointData As StandardPointData)

        If Me.SlotsAvailable > 0 Then
            ' We have enough space to write the provided point data to the data block.
            'm_fileStream.Seek(m_writeCursor - m_fileStream.Position, SeekOrigin.Current)   ' This is slower than
            m_fileStream.Seek(m_writeCursor, SeekOrigin.Begin)                              ' <--------------This
            m_fileStream.Write(pointData.BinaryImage, 0, StandardPointData.BinaryLength)
            m_writeCursor = m_fileStream.Position   ' Update the write cursor.
        Else
            Throw New InvalidOperationException("No slots available for writing new data.")
        End If

    End Sub

End Class

File: Source/OldCode/Archive/2005 TVA Code Library/Version/4.0/Source/TVA.Historian/Files/ArchiveDataBlock.cs
//*******************************************************************************************************
//  ArchiveDataBlock.cs
//  Copyright © 2009 - TVA, all rights reserved - Gbtc
//
//  Build Environment: C#, Visual Studio 2008
//  Primary Developer: Pinal C. Patel
//      Office: INFO SVCS APP DEV, CHATTANOOGA - MR BK-C
//       Phone: 423/751-3024
//       Email: pcpatel@tva.gov
//
//  Code Modification History:
//  -----------------------------------------------------------------------------------------------------
//  02/24/2007 - Pinal C. Patel
//       Generated original version of source code.
//  01/23/2008 - Pinal C. Patel
//       Removed IsForHistoricData and added IsActive to keep track of activity.
//  03/31/2008 - Pinal C. Patel
//       Modified code to use the same FileStream object used by FAT instead to creating a new one.
//       Removed IDisposable interface implementation and Size property.
//  04/20/2009 - Pinal C. Patel
//       Converted to C#.
//
//*******************************************************************************************************

using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;

namespace TVA.Historian.Files
{
    /// <summary>
    /// Represents a block of <see cref="ArchiveData"/> in an <see cref="ArchiveFile"/>.
    /// </summary>
    /// <seealso cref="ArchiveData"/>
    /// <seealso cref="ArchiveFile"/>
    public class ArchiveDataBlock
    {
        #region [ Members ]

        // Constants

        /// <summary>
        /// Time in seconds after which the block is considered inactive if no reads or writes were performed.
        /// </summary>
        private const int InactivityPeriod = 300;

        // Fields
        private int m_index;
        private int m_historianID;
        private ArchiveFile m_parent;
        private long m_writeCursor;
        private DateTime m_lastActivityTime;

        #endregion

        #region [ Constructors ]

        /// <summary>
        /// Initializes a new instance of the <see cref="ArchiveDataBlock"/> class. 
        /// </summary>
        /// <param name="parent">An <see cref="ArchiveFile"/> object.</param>
        /// <param name="index">0-based index of the <see cref="ArchiveDataBlock"/>.</param>
        /// <param name="historianID">Historian identifier whose <see cref="ArchiveData"/> is stored in the <see cref="ArchiveDataBlock"/>.</param>
        internal ArchiveDataBlock(ArchiveFile parent, int index, int historianID)
        {
            m_parent = parent;
            m_index = index;
            m_historianID = historianID;
            m_writeCursor = Location;
            m_lastActivityTime = DateTime.Now;
            // Initialize the write cursor position by reading existing data. We have to iterate through the returned
            // data in order for the write cursor to progress since Read() implements a "yield return" for return data.
            foreach (ArchiveData dataPoint in Read()) { }
        }

        #endregion

        #region [ Properties ]

        /// <summary>
        /// Gets the 0-based index of the <see cref="ArchiveDataBlock"/>.
        /// </summary>
        public int Index
        {
            get
            {
                return m_index;
            }
        }

        /// <summary>
        /// Gets the start location (byte position) of the <see cref="ArchiveDataBlock"/> in the <see cref="ArchiveFile"/>.
        /// </summary>
        public long Location
        {
            get
            {
                return (m_index * (m_parent.DataBlockSize * 1024));
            }
        }

        /// <summary>
        /// Gets the maximum number of <see cref="ArchiveData"/> points that can be stored in the <see cref="ArchiveDataBlock"/>.
        /// </summary>
        public int Capacity
        {
            get
            {
                return ((m_parent.DataBlockSize * 1024) / ArchiveData.ByteCount);
            }
        }

        /// <summary>
        /// Gets the number of <see cref="ArchiveData"/> points that have been written to the <see cref="ArchiveDataBlock"/>.
        /// </summary>
        public int SlotsUsed
        {
            get
            {
                return (int)((m_writeCursor - Location) / ArchiveData.ByteCount);
            }
        }

        /// <summary>
        /// Gets the number of <see cref="ArchiveData"/> points that can to written to the <see cref="ArchiveDataBlock"/>.
        /// </summary>
        public int SlotsAvailable
        {
            get
            {
                return (Capacity - SlotsUsed);
            }
        }

        /// <summary>
        /// Gets a boolean value that indicates whether the <see cref="ArchiveDataBlock"/> is being actively used.
        /// </summary>
        public bool IsActive
        {
            get
            {
                double inactivity = DateTime.Now.Subtract(m_lastActivityTime).TotalSeconds;
                if (inactivity <= InactivityPeriod)
                {
                    return true;
                }
                else
                {
                    Trace.WriteLine(string.Format("Inactive for {0} seconds (Last activity = {1}; Time now = {2})", inactivity, m_lastActivityTime, DateTime.Now));
                    return false;
                }
            }
        }

        #endregion

        #region [ Methods ]

        /// <summary>
        /// Reads existing <see cref="ArchiveData"/> points from the <see cref="ArchiveDataBlock"/>.
        /// </summary>
        public IEnumerable<ArchiveData> Read()
        {
            lock (m_parent.FileData)
            {
                // We'll start reading from where the data block begins.
                m_parent.FileData.Seek(Location, SeekOrigin.Begin);

                byte[] binaryImage = new byte[ArchiveData.ByteCount];
                for (int i = 1; i <= Capacity; i++)
                {
                    // Read the data in the block.
                    m_lastActivityTime = DateTime.Now;
                    m_parent.FileData.Read(binaryImage, 0, binaryImage.Length);
                    ArchiveData dataPoint = new ArchiveData(m_historianID, binaryImage, 0, binaryImage.Length);
                    if (!dataPoint.IsEmpty)
                    {
                        // There is data - use it.
                        m_writeCursor = m_parent.FileData.Position;
                        yield return dataPoint;
                    }
                    else
                    {
                        // Data is empty - stop reading.
                        yield break;
                    }
                }
            }
        }

        /// <summary>
        /// Writes the <paramref name="dataPoint"/> to the <see cref="ArchiveDataBlock"/>.
        /// </summary>
        /// <param name="dataPoint"><see cref="ArchiveData"/> point to write.</param>
        public void Write(ArchiveData dataPoint)
        {
            if (SlotsAvailable > 0)
            {
                // We have enough space to write the provided point data to the data block.
                m_lastActivityTime = DateTime.Now;
                lock (m_parent.FileData)
                {
                    // Write the data.
                    m_parent.FileData.Seek(m_writeCursor, SeekOrigin.Begin);
                    m_parent.FileData.Write(dataPoint.BinaryImage, 0, ArchiveData.ByteCount);
                    // Update the write cursor.
                    m_writeCursor = m_parent.FileData.Position;
                    // Flush the data if configured.
                    if (!m_parent.CacheWrites)
                        m_parent.FileData.Flush();
                }
            }
            else
            {
                throw (new InvalidOperationException("No slots available for writing new data."));
            }
        }

        /// <summary>
        /// Resets the <see cref="ArchiveDataBlock"/> by overwriting existing <see cref="ArchiveData"/> points with empty <see cref="ArchiveData"/> points.
        /// </summary>
        public void Reset()
        {
            m_writeCursor = Location;
            for (int i = 1; i <= Capacity; i++)
            {
                Write(new ArchiveData(m_historianID));
            }
            m_writeCursor = Location;
        }

        #endregion
    }
}